minibot.py

```diff
@@ -0,0 +1,272 @@
+                "Joystick index is out of range, should be 0-%s" % 2)
+
+        key = "/joystick-" + str(stick) + "/axis-" + str(axis)
+
+        try:
+            return self.table.getNumber(
+                key) if networktables.NetworkTables.isConnected() else 0.0
+        except KeyError:
+            # Joystick is unavailable, user should be notified?
+            return 0.0
+
+    def getStickButton(self, stick, button):
+        """The state of a button on the joystick. Button indexes begin at 1.
+        :param stick: The joystick port number
+        :type stick: int
+        :param button: The button index, beginning at 1.
+        :type button: int
+
+        :returns: The state of the button.
+        """
+        if stick < 0 or stick >= 2:
+            raise IndexError(
+                "Joystick index is out of range, should be 0-%s" % 2)
+
+        key = "/joystick-" + str(stick) + "/button-" + str(button)
+
+        try:
+            return self.table.getNumber(
+                key) != 0.0 if networktables.NetworkTables.isConnected(
+                ) else 0.0
+        except KeyError:
+            # Joystick is unavailable, user should be notified?
+            return 0.0
+
+    def isEnabled(self):
+        try:
+            return self.table.getBoolean("/enabled")
+        except KeyError:
+            # DS is not sending enabled/disabled
+            return False
+
+    def isDisabled(self):
+        try:
+            return not self.table.getBoolean("/enabled")
+        except KeyError:
+            # DS is not sending enabled/disabled
+            return True
+
+    def isAutonomous(self):
+        try:
+            return self.table.getString("/mode") == "autonomous"
+        except KeyError:
+            # DS is not sending mode
+            return False
+
+    def isOperatorControl(self):
+        try:
+            return self.table.getString("/mode") == "teleop"
+        except KeyError:
+            # DS is not sending mode
+            return False
+
+    def isTest(self):
+        try:
+            return self.table.getString("/mode") == "test"
+        except KeyError:
+            # DS is not sending mode
+            return False
+
+
+class Joystick:
+    """Joystick for Rpi
+
+    Handle input from standard Joysticks connected to the driverstation.
+
+    This class handles standard input that comes from the driverstation. Each
+    time a value is requested the most recent value is returned. There is a
+    single class instance for each joystick and the mapping of ports to
+    hardware buttons depends on the code in the drivestation.
+    """
+
+    def __init__(self, port):
+        """Construct an instance of a joystick.
+        The joystick index is the USB port on the Driver Station.
+
+        :param port: The port on the Driver Station that the joystick is
+            plugged into.
+        :type  port: int
+        """
+        self.port = port
+
+        # Class-instance of driver station
```

Comment:
I'm not a fan of using class-instances like this, because it makes it hard to reason about when & where the DriverStation gets created and torn down. If its a style change from the underlying robotpy stuff that this code is derived from, we can keep it for now, but I don't think this is a good design in general. (the robotpy design has a lot of stuff thats ugly because they did a very literal translation of the java stuff)

Reply:
It is pretty messy. I would like to experiment with creating a single instance of the `DriverStation` for each object that needs to access it. I mostly did it the way it is now to match up with robotpy, but I think that it's probably unnecessary.